### ISSUE TITLE
Implement active learning algorithms

### ISSUE THREAD

**Issue by MatthewJA** [chengsoonong/crowdastro#76]
- uncertainty sampling
- query-by-committee
- margin

**Comment by MatthewJA** [comment]
@chengsoonong: When training a new classifier after being given an oracle label, should this classifier _not_ be trained using a warm start? Or should the initial weights used be the old classifier weights? I've taken a look around but I haven't found anything about this.

**Comment by chengsoonong** [comment]
Warm start: This means that we initialise the weights of a classifier to something, before we train it. In the case of active learning, this should be the weights of the classifier from the previous epoch of active learning.

Logistic regression has a convex objective function, which means that no matter where you start the weights, you always end up in the same global minimum. Therefore there is no difference the resulting end classifier no matter how you initialise it, given the same set of training data. However, the hope is that we can save computation time by using warm start.

**Comment by MatthewJA** [comment]
So I should use a warm start. Can do!

**Comment by MatthewJA** [comment]
![uncertainty_random_active](https://cloud.githubusercontent.com/assets/2681312/15967341/987fb3c6-2f6a-11e6-932e-c8d8dc0ce772.png)

Uncertainty sampling vs. random sampling on a contrived `sklearn` dataset, averaged over 50 trials.

**Comment by MatthewJA** [comment]
I've implemented a QBC strategy (which does pretty poorly on my test data) and I'm working toward running uncertainty sampling with entropy on the RGZ data.

**Comment by MatthewJA** [comment]
Passive learning curve: #93

**Comment by MatthewJA** [comment]
![image](https://cloud.githubusercontent.com/assets/2681312/16332250/69c7487c-3a38-11e6-973b-9f63b1493b11.png)

Naïvely applying uncertainty sampling with entropy-based uncertainty.

**Comment by MatthewJA** [comment]
![image](https://cloud.githubusercontent.com/assets/2681312/16333881/0b58af02-3a41-11e6-8927-2838e2a065bb.png)

Same experiment, but also including the non-and/or'd data. Slightly different looking since I forgot to set the seed and it's initialised at random, but it's pretty clear that this doesn't work very well. I suspect entropy is the problem and I need to find a better way to quantify uncertainty for a subject.

**Comment by chengsoonong** [comment]
Get used to thinking about "random" as "passive". Change legends. The word random has too many meanings.

As I said right from the start, passive learning is pretty hard to beat.

**Comment by MatthewJA** [comment]
Sure, will do.

![image](https://cloud.githubusercontent.com/assets/2681312/16337749/18a269f4-3a5b-11e6-8ec1-33966306d1f7.png)

I want to find out whether changing the uncertainty collation improves results. I'm also curious as to whether FIRST benefits more from active learning.

**Comment by chengsoonong** [comment]
Regarding QBC:
Train 10 classifiers on about half of the total training set (i.e. your committee). Report the performance. If the performance is too similar or too different, then there might be some issues with the diversity of your committee.

**Comment by MatthewJA** [comment]
I'm pretty happy with all of this, so I'm going to close this issue. I'll probably make new issues for the active learning experiments discussed today.
